### ISSUE TITLE
[IDEA] Affordance to close/toggle a modal from outside the modal

### ISSUE THREAD

**Issue by saqimtiaz** [Jermolene/TiddlyWiki5#5831]
Currently if you open a modal with a global keyboard shortcut, there is no means to close it via another (or the same) global keyboard shortcut. The only means to close the modal is the `tm-close-tiddler` message sent from within the modal.

Quick ideas for how this might be implemented:

- user provided optional modal "name" parameter when opening a modal. Then a `tm-modal-close` message providing the same name can close that modal. OR,
- optional state tiddlers for each modal that is opened, where deleting the state tiddler closes the modal.

My use case: when a specific key combinaton is held down for 2 seconds, show a fullscreen modal styled as a semi-transparent overlay showing keyboard shortcuts, then hide it when the key is let go (`keyup` event and `eventcatcher`).

**Comment by Jermolene** [comment]
Interesting case. My first thought would be to not use a modal at all, but rather a `$:/tags/PageTemplate` tiddler containing a fixed position fake modal that you can show or hide via a list widget expression. You may be able to reuse the existing modal classes.

**Comment by saqimtiaz** [comment]
> Interesting case. My first thought would be to not use a modal at all, but rather a `$:/tags/PageTemplate` tiddler containing a fixed position fake modal that you can show or hide via a list widget expression. You may be able to reuse the existing modal classes.

I will probably end up going in that direction as I also want more control over the animations which are somewhat hardcoded for modals. However I think making the modal triggering/closing mechanism more flexible would be beneficial in other use cases as well.

**Comment by Jermolene** [comment]
As we've discussed previously, I am interested in upgrading the modal mechanism so that it can be used for dropdowns, overlays etc

**Comment by saqimtiaz** [comment]
@Jermolene I was thinking of that when creating this issue, so that we this use case in mind when we work on modals.

**Comment by yaisog** [comment]
My vote is on the state tiddler for a modal, where the state tiddler name is passed to the modal via parameter. Looking at how `mask-closable` was implemented, the state tiddler might also be passed via a field to the modal, at least until the modal implementation is upgraded.

A use case might be to check if the modal is already open when a key combination is used to open the modal, to avoid accidental double activations.

I have a local implementation running and would be able to make a PR for at least that intermediate step. Closing the modal by deleting the state tiddler would not be part of it (for now), though I do like the idea.

**Comment by Jermolene** [comment]
Hi @yaisog I think we're at the point where it's better to invest in refactoring the modal implementation into wikitext than making deep changes to the current system. There will be more backwards compatibility constraints if we add further features and semantics.

**Comment by yaisog** [comment]
Yeah, I didn't hold my breath for that one, because the backwards compatibility issues are clear. But let's not forget this when the modal implementation is upgraded...

**Comment by Jermolene** [comment]
> Yeah, I didn't hold my breath for that one, because the backwards compatibility issues are clear. But let's not forget this when the modal implementation is upgraded...

Perhaps it might be something you might like to work on? It would be an interesting area to pick up because it needs some exploration/experimentation, and plenty of consultation with the community to establish what people would like to see...

**Comment by yaisog** [comment]
> Perhaps it might be something you might like to work on?

I'm flattered, but I don't quite know what direction to start in. What did you have in mind for the modal overhaul?

I could see integrating it into the page template, as a top element controlled by a `$reveal` (which automatically solves the problem of implementing a state tiddler?). That would contain a `$list` to stack multiple modals.

You'd want it to be fully backwards-compatible, right? Which would have the neat benefit of shutting me up in the future about all kinds of feature requests... 😜 

One of the bigger problems would be the CSS, with all the animations and all. A lot of the code in modal.js seems to be concerned with those...
